contracts/multiply/GuniMultiplyProxyActions.sol

```diff
@@ -0,0 +1,422 @@
+    );
+
+    IManager(guniAddressRegistry.manager).cdpAllow(
+      cdpData.cdpId,
+      guniAddressRegistry.guniProxyActions,
+      0
+    );
+  }
+
+  function _increaseMPGuni(
+    ExchangeData memory exchangeData,
+    CdpData memory cdpData,
+    GuniAddressRegistry memory guniAddressRegistry,
+    uint256 borrowedDaiAmount
+  ) internal {
+    IGUNIToken guni = IGUNIToken(guniAddressRegistry.guni);
+    IERC20 otherToken = IERC20(guniAddressRegistry.otherToken);
+
+    uint256 bal0 = daiContract.balanceOf(address(this));
+
+    {
+      IExchange exchange = IExchange(guniAddressRegistry.exchange);
+
+      daiContract.approve(address(exchange), exchangeData.fromTokenAmount);
+
+      exchange.swapDaiForToken(
+        exchangeData.toTokenAddress,
+        exchangeData.fromTokenAmount,
+        exchangeData.minToTokenAmount,
+        exchangeData.exchangeAddress,
+        exchangeData._exchangeCalldata
+      );
+    }
+
+    uint256 guniBalance;
+    uint256 bal1 = otherToken.balanceOf(address(this)); //120k
+    bal0 = daiContract.balanceOf(address(this)); //80 k
+
+    {
+      IGUNIRouter router = IGUNIRouter(guniAddressRegistry.router);
+      daiContract.approve(address(router), bal0);
+      otherToken.approve(address(router), bal1);
+
+      (, , guniBalance) = router.addLiquidity(address(guni), bal0, bal1, 0, 0, address(this));
+    }
+
+    guni.approve(guniAddressRegistry.guniProxyActions, guniBalance);
+    joinDrawDebt(cdpData, borrowedDaiAmount, guniAddressRegistry.manager, guniAddressRegistry.jug);
+
+    uint256 daiLeft = IERC20(DAI).balanceOf(address(this)).sub(borrowedDaiAmount);
+    uint256 otherTokenLeft = otherToken.balanceOf(address(this));
+
+    if (daiLeft > 0) {
+      IERC20(DAI).transfer(cdpData.fundsReceiver, daiLeft);
+    }
+    if (otherTokenLeft > 0) {
+      otherToken.transfer(cdpData.fundsReceiver, otherTokenLeft);
+    }
+  }
+
+  function _closeToDaiMPGuni(
+    ExchangeData memory exchangeData,
+    CdpData memory cdpData,
+    GuniAddressRegistry memory guniAddressRegistry,
+    uint256 borrowedDaiAmount
+  ) internal {
+    IExchange exchange = IExchange(guniAddressRegistry.exchange);
+    IERC20 otherToken = IERC20(guniAddressRegistry.otherToken);
+    uint256 ink = getInk(guniAddressRegistry.manager, cdpData);
+
+    wipeAndFreeGem(
+      guniAddressRegistry.manager,
+      cdpData.gemJoin,
+      cdpData.cdpId,
+      cdpData.requiredDebt,
+      ink
+    );
+
+    IGUNIToken guni = IGUNIToken(guniAddressRegistry.guni);
+
+    uint256 guniBalance = guni.balanceOf(address(this));
+
+    {
+      IGUNIRouter router = IGUNIRouter(guniAddressRegistry.router);
+      guni.approve(address(router), guniBalance);
+      router.removeLiquidity(address(guni), guniBalance, 0, 0, address(this));
+    }
+
+    otherToken.approve(address(exchange), otherToken.balanceOf(address(this)));
+    exchange.swapTokenForDai(
+      exchangeData.toTokenAddress,
+      otherToken.balanceOf(address(this)),
```

Comment:
why this value do not come from front-end ? what if it does not match calldata ?